odl/contrib/solvers/spdhg/misc.py

```diff
@@ -0,0 +1,857 @@
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the total variation of a constant is zero
+
+        >>> import odl.contrib.spdhg as spdhg, odl
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = space.one()
+        >>> TVNN(x) < 1e-10
+
+        Check that negative functions are mapped to infty
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> np.isinf(TVNN(x))
+        """
+
+        nn = self.NN(x)
+
+        if nn is np.inf:
+            return nn
+        else:
+            out = self.alpha * self.TV(x) + nn
+            if self.strong_convexity > 0:
+                out += self.strong_convexity * self.L2(x)
+            return out
+
+    def proximal(self, sigma):
+        """ Prox operator of TV. It allows the proximal step length to be a vector
+        of positive elements.
+
+        Parameters
+        ----------
+        x : np.array
+            vector / image
+
+        Returns
+        -------
+        extended float (with infinity)
+            Is the input in the non-negative orthant?
+
+        Examples
+        --------
+        Check that the proximal operator is the identity for sigma=0
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> y = TVNN.proximal(0)(x)
+        >>> (y-x).norm() < 1e-10
+
+        Check that negative functions are mapped to 0
+
+        >>> import odl.contrib.spdhg as spdhg, odl, numpy as np
+        >>> space = odl.uniform_discr([0, 0], [3, 3], [3, 3])
+        >>> TVNN = spdhg.TV_NonNegative(space, alpha=2)
+        >>> x = -space.one()
+        >>> y = TVNN.proximal(0.1)(x)
+        >>> y.norm() < 1e-10
+        """
+
+        if sigma == 0:
+            return odl.IdentityOperator(self.domain)
+
+        else:
+            def TV_prox(z, out=None):
+
+                if out is None:
+                    out = z.space.zero()
+
+                opts = self.prox_options
+
+                sigma_ = np.copy(sigma)
+                z_ = z.copy()
+
+                if self.strong_convexity > 0:
+                    sigma_ /= (1 + sigma * self.strong_convexity)
+                    z_ /= (1 + sigma * self.strong_convexity)
+
+                def proj_C(x, out=None):
+                    return self.NN.proximal(1)(x, out)
+
+                def proj_P(x, out=None):
+                    return odl.solvers.GroupL1Norm(
+                            self.grad.range, exponent=2
+                            ).convex_conj.proximal(0)(x, out=out)
```

Comment:
This is a bit ugly. Better to assign the norm to a variable and use `convex_conj...` on that one. It will improve readability.